components/camera/rtsp/decode.go

```diff
@@ -2,26 +2,146 @@ package rtsp
+	}
+
+	params, err := dec.fmtCxt.CodecParameters(ctx, dec.videoStream)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get codec parameters")
+	}
+
+	id, err := params.CodecID(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get codec ID from parameters")
+	}
+
+	if dec.codec, err = ffmpeg.FindDecoder(ctx, id); err != nil {
+		return nil, errors.Wrap(err, "cannot find decoder")
+	}
+
+	if dec.context, err = dec.codec.AllocContext3(ctx); err != nil {
+		return nil, errors.Wrap(err, "cannot allocate context from codec")
+	}
+
+	if err := dec.fmtCxt.ParamsToContext(ctx, dec.context, dec.videoStream); err != nil {
+		return nil, errors.Wrap(err, "cannot convert parameters to context")
+	}
+
+	if err := dec.context.Open2(ctx, dec.codec); err != nil {
+		return nil, errors.Wrap(err, "cannot open decoder")
+	}
+
+	dec.fmtCxt.ReadPlay(ctx)
+	dec.context.FlushBuffers()
+
+	return dec, nil
+}
+
+// Close closes the decoder.
+func (dec *Decoder) Close() {
+	dec.cancelFunc()
+	dec.context.Close()
+	dec.fmtCxt.Close()
+}
+
+// Decode retrieves and (optionally for H264) decodes the next frame returned from the RTSP stream.
+func (dec *Decoder) Decode(ctx context.Context) (image.Image, error) {
+	// See https://ffmpeg.org/doxygen/trunk/api-band-test_8c_source.html#l00163
+	frame, err := ffmpeg.FrameAlloc()
+	if err != nil {
+		return nil, err
+	}
+	defer frame.Free()
+
+	pkt, err := ffmpeg.PacketAlloc()
+	if err != nil {
+		return nil, err
+	}
+	defer pkt.Free()
+
+	for {
+		if err := dec.fmtCxt.ReadFrame(ctx, pkt); err == nil && pkt.StreamIndex() != dec.videoStream {
+			pkt.Unref()
+			continue
+		}
+
+		switch pixFmt := dec.context.PixFmt(); pixFmt {
+		case ffmpeg.PixFmtYUV420P, ffmpeg.PixFmtYUV422P, ffmpeg.PixFmtYUV444P:
+			result := dec.context.SendPacket(pkt)
+			pkt.Unref()
+			if result < 0 {
+				return nil, errors.Wrap(ffmpeg.ErrorFromCode(result), "cannot send packet")
+			}
+
+			if result = dec.context.ReceiveFrame(ctx, frame); result == ffmpeg.EOF {
+				return nil, errors.Wrap(ffmpeg.ErrorFromCode(result), "cannot receive frame")
+			} else if result == ffmpeg.EAGAIN {
+				continue
+			} else if result < 0 {
+				return nil, errors.Wrap(ffmpeg.ErrorFromCode(result), "error decoding frame")
+			}
 
-func mjpegDecoding() (*format.MJPEG, decoder) {
-	var mjpeg format.MJPEG
-	// get the RTP->MJPEG decoder
-	rtpDec := mjpeg.CreateDecoder()
-	mjpegDecoder := func(pkt *rtp.Packet) (image.Image, error) {
-		encoded, _, err := rtpDec.Decode(pkt)
-		if err != nil {
-			return nil, errors.Wrap(err, "rtp to mjpeg decoding failed")
+			img := frame.ToImage()
+			frame.Unref()
+			return img, nil
+		case ffmpeg.PixFmtYUVJ420P, ffmpeg.PixFmtYUVJ422P, ffmpeg.PixFmtYUVJ444P:
+			b := bytes.Buffer{}
+			b.Write(unsafe.Slice(pkt.Data(), pkt.Size())) // we must copy bytes before pkt.Unref or pkt.Free
+			// TODO: ideally, we'd do something similar for H264, storing the raw bytes in a lazy image and decoding
+			// them client-side.
```

Comment:
Trying to pass through the H264 bytes kept failing. For some reason, the frontend couldn't decode the raw bytes.

Reply:
[q] What datastructure were you using for h264 passthrough attempts? Wondering if we could use `NewLazyEncodedImage` with custom mime-type.